lib/Dialect/TritonGPU/IR/Dialect.cpp

```diff
@@ -147,6 +147,20 @@ SmallVector<unsigned> getShapePerCTATile(Attribute layout) {
 }
 
 bool isExpensiveView(Type srcType, Type dstType) {
+  auto tensorSrcType = cast<RankedTensorType>(srcType);
+  auto tensorDstType = cast<RankedTensorType>(dstType);
+  auto llSrc =
+      toLinearLayout(tensorSrcType.getShape(), tensorSrcType.getEncoding());
+  auto llDst =
+      toLinearLayout(tensorDstType.getShape(), tensorDstType.getEncoding());
+  // In case there are replicated value we need to make sure the new and old
+  // layout have matching masks.
+  for (auto [srcMask, dstMask] :
+       llvm::zip(llSrc.getFreeVariableMasks(), llDst.getFreeVariableMasks())) {
+    assert(srcMask.first == dstMask.first);
+    if (srcMask.second != dstMask.second)
+      return true;
+  }
   return getTotalElemsPerThread(srcType) != getTotalElemsPerThread(dstType);
```

Comment:
I guess this is not needed anymore, right?

Reply:
What if the free masks are 1000 vs 10000?

Reply:
the register mask should have the number of elements baked into it via the number of zeros it has, so if the masks are equal, the number of elements should be equal.

Reply:
right, so `getTotalElemsPerThread` would always match at this point? Good point

Reply:
Hmm, do we have it right now? 

From the implementation I don't think so

Reply:
in what case would the first condition pass and `getTotalElemsPerThread` would return different values?

Reply:
It's the case I mentioned, we have free masks L1: 1000 vs L2: 10000. L1 has 16 registers but L2 has 32 registers.
Though I'm not sure if in practice you are able to trigger it.
If we do have the number of elements in the mask indeed we don't have to check the last condition but my understanding is that currently we do not.

Reply:
but with those masks, the condition Thomas added would already return `true`? My point is that if all the masks are equal (in particular the one for registers) then we know that the layouts have the same number of registers.

Reply:
@Jokeren has a point. This is a free function, and what I said is correct provided that the two inputs represent the same tensor, as such, we could do that by asserting that the shapes and elements of the RankedTensorTypes are indeed the same, but yeah.

Reply:
only the total number of elements in the tensor need to match, not the shape right?

Reply:
yep, that's right